[CIR] `__sync_*` compare-and-swap, swap and lock builtins unimplemented; NYI diagnostic mislabels them as "X86 builtin call"

[CIR] `__sync_*` compare-and-swap, swap and lock builtins unimplemented; NYI diagnostic mislabels them as "X86 builtin call"

## Summary

Two related issues in CIRGen's handling of the legacy GCC `__sync_*` atomic
builtins:

1. **Five `__sync_*` builtins are unimplemented.**
   `__sync_bool_compare_and_swap`, `__sync_val_compare_and_swap`,
   `__sync_swap`, `__sync_lock_test_and_set` and `__sync_lock_release` all emit
   `error: ClangIR code gen Not Yet Implemented`. The other thirteen members of
   the same family are implemented and work, so support is currently partial.
2. **The diagnostic says "X86 builtin call" for builtins that are not
   X86-specific.** These are generic GCC builtins available on every target,
   and the same message appears verbatim when targeting aarch64. The helper
   that emits it lives in the target-independent `CIRGenBuiltin.cpp` but
   hardcodes "X86", so it mislabels every builtin routed through it.

Issue 2 is cosmetic, but it actively misleads triage — it suggests a
target-specific problem where none exists. It is a two-line fix, independent of
issue 1.

## Reduced test case

```c
// sync.c
int  cas_bool(int *p)     { return __sync_bool_compare_and_swap(p, 0, 1); }
int  cas_val(int *p)      { return __sync_val_compare_and_swap(p, 0, 1); }
int  swap(int *p)         { return __sync_swap(p, 1); }
int  test_and_set(int *p) { return __sync_lock_test_and_set(p, 1); }
void lock_release(int *p) { __sync_lock_release(p); }
```

```console
$ clang -fclangir -emit-cir -c sync.c -o /dev/null
sync.c:2:36: error: ClangIR code gen Not Yet Implemented: unimplemented X86 builtin call: __sync_bool_compare_and_swap_4
    2 | int  cas_bool(int *p)     { return __sync_bool_compare_and_swap(p, 0, 1); }
      |                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

No headers required. Compilation stops at the first error, so each of the five
must be compiled separately to see its own diagnostic; all five behave the same
way.

The result is identical for every operand width (1, 2, 4, 8 bytes) and at every
optimization level from `-O0` to `-O3`.

### Mislabelled diagnostic on a non-x86 target

```console
$ clang -fclangir -emit-cir --target=aarch64-linux-gnu -c sync.c -o /dev/null
sync.c:2:36: error: ClangIR code gen Not Yet Implemented: unimplemented X86 builtin call: __sync_bool_compare_and_swap_4
                                                                       ^^^ targeting aarch64
```

## Component triage

Per [How to submit an LLVM bug report](https://llvm.org/docs/HowToSubmitABug.html),
this is a **front-end (CIRGen)** issue. The classic pipeline compiles the same
file without complaint:

| Command | Result |
|---|---|
| `clang -emit-llvm -Xclang -disable-llvm-passes -O2 -c sync.c` | ok |
| `clang -emit-llvm -O2 -c sync.c` | ok |
| `clang -O2 -c sync.c` | ok |
| `clang -fclangir -emit-cir -O2 -c sync.c` | **NYI error** |

Classic CodeGen lowers these to ordinary atomic IR. For `cas_bool`:

```llvm
%2 = cmpxchg ptr %0, i32 0, i32 1 seq_cst seq_cst, align 4
```

This is a missing-feature report, not a crash — codegen stops cleanly at the
diagnostic.

## Which `__sync_*` builtins work

Support is partial. The arithmetic operations are implemented; the
compare-and-swap, swap and lock operations are not.

| Builtin | Status |
|---|---|
| `__sync_fetch_and_add` / `_sub` / `_or` / `_and` / `_xor` / `_nand` | works |
| `__sync_add_and_fetch` / `_sub_and_fetch` / `_and_and_fetch` | works |
| `__sync_or_and_fetch` / `_xor_and_fetch` / `_nand_and_fetch` | works |
| `__sync_synchronize` | works |
| `__sync_bool_compare_and_swap` | **NYI** |
| `__sync_val_compare_and_swap` | **NYI** |
| `__sync_swap` | **NYI** |
| `__sync_lock_test_and_set` | **NYI** |
| `__sync_lock_release` | **NYI** |

For contrast, the modern `__atomic_*` equivalents are all implemented and
compile fine, including `__atomic_compare_exchange_n`, `__atomic_exchange_n`,
`__atomic_store_n` and `__atomic_fetch_add`. Only the legacy spellings are
missing, so this affects codebases that predate C11 atomics or deliberately
target older compilers.

## Where this happens

In `clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp`, the arithmetic cases dispatch to
a real implementation:

```c++
  case Builtin::BI__sync_fetch_and_add_1:
  ...
  case Builtin::BI__sync_fetch_and_add_16:
    return emitBinaryAtomic(*this, cir::AtomicFetchKind::Add, e);
```

while the CAS, swap and lock cases fall through to the NYI helper:

```c++
  case Builtin::BI__sync_val_compare_and_swap_1:
  ...
  case Builtin::BI__sync_lock_release_16:
    return errorBuiltinNYI(*this, e, builtinID);
```

Both the unsuffixed forms (`BI__sync_bool_compare_and_swap`) and the
size-suffixed forms (`BI__sync_bool_compare_and_swap_1` … `_16`) route to
`errorBuiltinNYI`.

Since `emitBinaryAtomic` and `emitBinaryAtomicPost` already exist for the
arithmetic cases, and CIR already supports `__atomic_compare_exchange_n` and
`__atomic_exchange_n`, implementing these should largely be a matter of mapping
the legacy builtins onto machinery that is already present.

### The "X86" label

`errorBuiltinNYI` is defined in the target-independent `CIRGenBuiltin.cpp` and
unconditionally says "X86":

```c++
static RValue errorBuiltinNYI(CIRGenFunction &cgf, const CallExpr *e,
                              unsigned builtinID) {
  if (cgf.getContext().BuiltinInfo.isLibFunction(builtinID)) {
    cgf.cgm.errorNYI(
        e->getSourceRange(),
        std::string("unimplemented X86 library function builtin call: ") +
            cgf.getContext().BuiltinInfo.getName(builtinID));
  } else {
    cgf.cgm.errorNYI(e->getSourceRange(),
                     std::string("unimplemented X86 builtin call: ") +
                         cgf.getContext().BuiltinInfo.getName(builtinID));
  }
  ...
```

There are 20 call sites of this helper in that file, all for target-independent
builtins. `__builtin_thread_pointer` is another one affected: it also reports
"unimplemented X86 builtin call" when targeting aarch64. Dropping `"X86 "` from
both strings would fix it.

## Impact: GNU binutils

Building [GNU binutils](https://sourceware.org/git/binutils-gdb.git) with
ClangIR, this is the **only** thing that fails. Of 1122 translation units in an
`--enable-targets=all` build, 1116 compile. The 6 failures are three source
files in the vendored `libbacktrace`, each compiled twice by libtool (PIC and
non-PIC):

| Source file | Line | Builtin |
|---|---|---|
| `libbacktrace/dwarf.c` | 4060 | `__sync_bool_compare_and_swap` |
| `libbacktrace/elf.c` | 7445 | `__sync_bool_compare_and_swap` |
| `libbacktrace/mmap.c` | 135 | `__sync_lock_test_and_set` |
| `libbacktrace/mmap.c` | 162 | `__sync_lock_release` |

`libbacktrace` uses the legacy spellings deliberately, for compatibility with
compilers predating C11 atomics.

To reproduce:

```bash
git clone https://sourceware.org/git/binutils-gdb.git
mkdir binutils-build && cd binutils-build
../binutils-gdb/configure --disable-gdb --disable-gdbserver --disable-sim \
                          --disable-gprofng --disable-werror --disable-nls
make                       # generates the config.h libbacktrace needs
cd libbacktrace            # the compile runs from this directory

for f in dwarf.c elf.c mmap.c; do
  path/to/clang -fclangir -emit-cir -O2 -DHAVE_CONFIG_H -I. \
    -I../../binutils-gdb/libbacktrace \
    -I../../binutils-gdb/libbacktrace/../include \
    -c ../../binutils-gdb/libbacktrace/$f -o /dev/null
done
```

## Environment

- **Source**: upstream `https://github.com/llvm/llvm-project`, commit `91b781417ecdf354afa0c80450b970db11d63130` (`main`)
- **Target**: `x86_64-unknown-linux-gnu` (also reproduced with `--target=aarch64-linux-gnu`)
- **Host**: Debian, gcc 15.2.0 as host compiler
- **Build**:
  ```bash
  cmake -G Ninja -S llvm -B build \
    -DCMAKE_BUILD_TYPE=Release \
    -DLLVM_ENABLE_PROJECTS="clang;mlir" \
    -DCLANG_ENABLE_CIR=ON
  ninja -C build clang
  ```

--- a/clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp
+++ b/clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp
@@ -247,6 +247,61 @@ static RValue emitBinaryAtomicPost(CIRGenFunction &cgf,
   return RValue::get(result);
 }
 
+/// Emit a `cir.atomic.cmpxchg` for __sync_val_compare_and_swap_N and
+/// __sync_bool_compare_and_swap_N. Returns the old value when `returnBool` is
+/// false, otherwise returns a boolean success flag.
+static RValue emitAtomicCmpXchg(CIRGenFunction &cgf, const CallExpr *e,
+                                bool returnBool) {
+  Address destAddr = checkAtomicAlignment(cgf, e);
+  CIRGenBuilderTy &builder = cgf.getBuilder();
+  mlir::Value destValue = destAddr.emitRawPointer();
+  mlir::Value expected = cgf.emitScalarExpr(e->getArg(1));
+  mlir::Value desired = cgf.emitScalarExpr(e->getArg(2));
+
+  auto cmpxchg = cir::AtomicCmpXchgOp::create(
+      builder, cgf.getLoc(e->getSourceRange()), destValue, expected, desired,
+      cir::MemOrder::SequentiallyConsistent,
+      cir::MemOrder::SequentiallyConsistent, cir::SyncScopeKind::System,
+      /*alignment=*/nullptr, /*weak=*/false, /*is_volatile=*/false);
+
+  if (returnBool) {
+    // cir.atomic.cmpxchg already returns (old, success). Use the success flag
+    // directly instead of re-emitting the expected argument and comparing.
+    return RValue::get(cmpxchg.getSuccess());
+  }
+  return RValue::get(cmpxchg.getOld());
+}
+
+/// Emit a `cir.atomic.xchg` for __sync_swap_N and __sync_lock_test_and_set_N.
+static RValue emitAtomicXchg(CIRGenFunction &cgf, const CallExpr *e,
+                             cir::MemOrder ordering) {
+  Address destAddr = checkAtomicAlignment(cgf, e);
+  CIRGenBuilderTy &builder = cgf.getBuilder();
+  mlir::Value destValue = destAddr.emitRawPointer();
+  mlir::Value val = cgf.emitScalarExpr(e->getArg(1));
+
+  auto xchg = cir::AtomicXchgOp::create(
+      builder, cgf.getLoc(e->getSourceRange()), destValue, val, ordering,
+      cir::SyncScopeKind::System, /*is_volatile=*/false);
+  return RValue::get(xchg.getResult());
+}
+
+/// Emit a release store of 0 for __sync_lock_release_N.
+static void emitAtomicLockRelease(CIRGenFunction &cgf, const CallExpr *e) {
+  CIRGenBuilderTy &builder = cgf.getBuilder();
+  Address destAddr = checkAtomicAlignment(cgf, e);
+  mlir::Location loc = cgf.getLoc(e->getSourceRange());
+  mlir::Type elemTy = destAddr.getElementType();
+  mlir::Value zero = builder.getConstant(loc, builder.getZeroInitAttr(elemTy));
+  auto orderAttr =
+      cir::MemOrderAttr::get(&cgf.getMLIRContext(), cir::MemOrder::Release);
+  auto scopeAttr = cir::SyncScopeKindAttr::get(&cgf.getMLIRContext(),
+                                               cir::SyncScopeKind::System);
+  builder.createStore(loc, zero, destAddr, /*isVolatile=*/false,
+                      /*isNontemporal=*/false,
+                      /*align=*/mlir::IntegerAttr{}, scopeAttr, orderAttr);
+}
+
 static void emitAtomicFenceOp(CIRGenFunction &cgf, const CallExpr *expr,
                               cir::SyncScopeKind syncScope) {
   CIRGenBuilderTy &builder = cgf.getBuilder();
@@ -432,11 +487,11 @@ static RValue errorBuiltinNYI(CIRGenFunction &cgf, const CallExpr *e,
   if (cgf.getContext().BuiltinInfo.isLibFunction(builtinID)) {
     cgf.cgm.errorNYI(
         e->getSourceRange(),
-        std::string("unimplemented X86 library function builtin call: ") +
+        std::string("unimplemented library function builtin call: ") +
             cgf.getContext().BuiltinInfo.getName(builtinID));
   } else {
     cgf.cgm.errorNYI(e->getSourceRange(),
-                     std::string("unimplemented X86 builtin call: ") +
+                     std::string("unimplemented builtin call: ") +
                          cgf.getContext().BuiltinInfo.getName(builtinID));
   }
 
@@ -2052,7 +2107,7 @@ RValue CIRGenFunction::emitBuiltinExpr(const GlobalDecl &gd, unsigned builtinID,
   case Builtin::BI__sync_lock_test_and_set:
   case Builtin::BI__sync_lock_release:
   case Builtin::BI__sync_swap:
-    return errorBuiltinNYI(*this, e, builtinID);
+    llvm_unreachable("Shouldn't make it through sema");
   case Builtin::BI__sync_fetch_and_add_1:
   case Builtin::BI__sync_fetch_and_add_2:
   case Builtin::BI__sync_fetch_and_add_4:
@@ -2140,26 +2195,32 @@ RValue CIRGenFunction::emitBuiltinExpr(const GlobalDecl &gd, unsigned builtinID,
   case Builtin::BI__sync_val_compare_and_swap_2:
   case Builtin::BI__sync_val_compare_and_swap_4:
   case Builtin::BI__sync_val_compare_and_swap_8:
-  case Builtin::BI__sync_val_compare_and_swap_16:
+    return emitAtomicCmpXchg(*this, e, /*returnBool=*/false);
   case Builtin::BI__sync_bool_compare_and_swap_1:
   case Builtin::BI__sync_bool_compare_and_swap_2:
   case Builtin::BI__sync_bool_compare_and_swap_4:
   case Builtin::BI__sync_bool_compare_and_swap_8:
-  case Builtin::BI__sync_bool_compare_and_swap_16:
+    return emitAtomicCmpXchg(*this, e, /*returnBool=*/true);
   case Builtin::BI__sync_swap_1:
   case Builtin::BI__sync_swap_2:
   case Builtin::BI__sync_swap_4:
   case Builtin::BI__sync_swap_8:
-  case Builtin::BI__sync_swap_16:
+    return emitAtomicXchg(*this, e, cir::MemOrder::SequentiallyConsistent);
   case Builtin::BI__sync_lock_test_and_set_1:
   case Builtin::BI__sync_lock_test_and_set_2:
   case Builtin::BI__sync_lock_test_and_set_4:
   case Builtin::BI__sync_lock_test_and_set_8:
-  case Builtin::BI__sync_lock_test_and_set_16:
+    return emitAtomicXchg(*this, e, cir::MemOrder::SequentiallyConsistent);
   case Builtin::BI__sync_lock_release_1:
   case Builtin::BI__sync_lock_release_2:
   case Builtin::BI__sync_lock_release_4:
   case Builtin::BI__sync_lock_release_8:
+    emitAtomicLockRelease(*this, e);
+    return RValue::get(nullptr);
+  case Builtin::BI__sync_val_compare_and_swap_16:
+  case Builtin::BI__sync_bool_compare_and_swap_16:
+  case Builtin::BI__sync_swap_16:
+  case Builtin::BI__sync_lock_test_and_set_16:
   case Builtin::BI__sync_lock_release_16:
     return errorBuiltinNYI(*this, e, builtinID);
   case Builtin::BI__sync_synchronize: {

--- a/clang/test/CIR/CodeGenBuiltins/builtin-sync.cpp
+++ b/clang/test/CIR/CodeGenBuiltins/builtin-sync.cpp
@@ -1,0 +1,186 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir -emit-cir %s -o %t.cir
+// RUN: FileCheck --check-prefix=CIR --input-file=%t.cir %s
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir -emit-llvm %s -o %t-cir.ll
+// RUN: FileCheck --check-prefix=LLVM --input-file=%t-cir.ll %s
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -emit-llvm %s -o %t.ll
+// RUN: FileCheck --check-prefix=LLVM --input-file=%t.ll %s
+
+extern "C" {
+
+// __sync_val_compare_and_swap
+
+// CIR-LABEL: @test_sync_val_compare_and_swap_1(
+// CIR: cir.atomic.cmpxchg success(seq_cst) failure(seq_cst) syncscope(system) %{{.+}}, %{{.+}}, %{{.+}} : (!cir.ptr<!s8i>, !s8i, !s8i) -> (!s8i, !cir.bool)
+// LLVM-LABEL: @test_sync_val_compare_and_swap_1(
+// LLVM: cmpxchg ptr %{{.+}}, i8 %{{.+}}, i8 %{{.+}} seq_cst seq_cst, align 1
+char test_sync_val_compare_and_swap_1(char *p, char oldv, char newv) {
+  return __sync_val_compare_and_swap(p, oldv, newv);
+}
+
+// CIR-LABEL: @test_sync_val_compare_and_swap_2(
+// CIR: cir.atomic.cmpxchg success(seq_cst) failure(seq_cst) syncscope(system) %{{.+}}, %{{.+}}, %{{.+}} : (!cir.ptr<!s16i>, !s16i, !s16i) -> (!s16i, !cir.bool)
+// LLVM-LABEL: @test_sync_val_compare_and_swap_2(
+// LLVM: cmpxchg ptr %{{.+}}, i16 %{{.+}}, i16 %{{.+}} seq_cst seq_cst, align 2
+short test_sync_val_compare_and_swap_2(short *p, short oldv, short newv) {
+  return __sync_val_compare_and_swap(p, oldv, newv);
+}
+
+// CIR-LABEL: @test_sync_val_compare_and_swap_4(
+// CIR: cir.atomic.cmpxchg success(seq_cst) failure(seq_cst) syncscope(system) %{{.+}}, %{{.+}}, %{{.+}} : (!cir.ptr<!s32i>, !s32i, !s32i) -> (!s32i, !cir.bool)
+// LLVM-LABEL: @test_sync_val_compare_and_swap_4(
+// LLVM: cmpxchg ptr %{{.+}}, i32 %{{.+}}, i32 %{{.+}} seq_cst seq_cst, align 4
+int test_sync_val_compare_and_swap_4(int *p, int oldv, int newv) {
+  return __sync_val_compare_and_swap(p, oldv, newv);
+}
+
+// CIR-LABEL: @test_sync_val_compare_and_swap_8(
+// CIR: cir.atomic.cmpxchg success(seq_cst) failure(seq_cst) syncscope(system) %{{.+}}, %{{.+}}, %{{.+}} : (!cir.ptr<!s64i>, !s64i, !s64i) -> (!s64i, !cir.bool)
+// LLVM-LABEL: @test_sync_val_compare_and_swap_8(
+// LLVM: cmpxchg ptr %{{.+}}, i64 %{{.+}}, i64 %{{.+}} seq_cst seq_cst, align 8
+long long test_sync_val_compare_and_swap_8(long long *p, long long oldv,
+                                           long long newv) {
+  return __sync_val_compare_and_swap(p, oldv, newv);
+}
+
+// __sync_bool_compare_and_swap
+
+// CIR-LABEL: @test_sync_bool_compare_and_swap_1(
+// CIR: cir.atomic.cmpxchg success(seq_cst) failure(seq_cst) syncscope(system) %{{.+}}, %{{.+}}, %{{.+}} : (!cir.ptr<!s8i>, !s8i, !s8i) -> (!s8i, !cir.bool)
+// LLVM-LABEL: @test_sync_bool_compare_and_swap_1(
+// LLVM: cmpxchg ptr %{{.+}}, i8 %{{.+}}, i8 %{{.+}} seq_cst seq_cst, align 1
+// LLVM: extractvalue { i8, i1 } %{{.+}}, 1
+bool test_sync_bool_compare_and_swap_1(char *p, char oldv, char newv) {
+  return __sync_bool_compare_and_swap(p, oldv, newv);
+}
+
+// CIR-LABEL: @test_sync_bool_compare_and_swap_2(
+// CIR: cir.atomic.cmpxchg success(seq_cst) failure(seq_cst) syncscope(system) %{{.+}}, %{{.+}}, %{{.+}} : (!cir.ptr<!s16i>, !s16i, !s16i) -> (!s16i, !cir.bool)
+// LLVM-LABEL: @test_sync_bool_compare_and_swap_2(
+// LLVM: cmpxchg ptr %{{.+}}, i16 %{{.+}}, i16 %{{.+}} seq_cst seq_cst, align 2
+// LLVM: extractvalue { i16, i1 } %{{.+}}, 1
+bool test_sync_bool_compare_and_swap_2(short *p, short oldv, short newv) {
+  return __sync_bool_compare_and_swap(p, oldv, newv);
+}
+
+// CIR-LABEL: @test_sync_bool_compare_and_swap_4(
+// CIR: cir.atomic.cmpxchg success(seq_cst) failure(seq_cst) syncscope(system) %{{.+}}, %{{.+}}, %{{.+}} : (!cir.ptr<!s32i>, !s32i, !s32i) -> (!s32i, !cir.bool)
+// LLVM-LABEL: @test_sync_bool_compare_and_swap_4(
+// LLVM: cmpxchg ptr %{{.+}}, i32 %{{.+}}, i32 %{{.+}} seq_cst seq_cst, align 4
+// LLVM: extractvalue { i32, i1 } %{{.+}}, 1
+bool test_sync_bool_compare_and_swap_4(int *p, int oldv, int newv) {
+  return __sync_bool_compare_and_swap(p, oldv, newv);
+}
+
+// CIR-LABEL: @test_sync_bool_compare_and_swap_8(
+// CIR: cir.atomic.cmpxchg success(seq_cst) failure(seq_cst) syncscope(system) %{{.+}}, %{{.+}}, %{{.+}} : (!cir.ptr<!s64i>, !s64i, !s64i) -> (!s64i, !cir.bool)
+// LLVM-LABEL: @test_sync_bool_compare_and_swap_8(
+// LLVM: cmpxchg ptr %{{.+}}, i64 %{{.+}}, i64 %{{.+}} seq_cst seq_cst, align 8
+// LLVM: extractvalue { i64, i1 } %{{.+}}, 1
+bool test_sync_bool_compare_and_swap_8(long long *p, long long oldv,
+                                       long long newv) {
+  return __sync_bool_compare_and_swap(p, oldv, newv);
+}
+
+// __sync_swap
+
+// CIR-LABEL: @test_sync_swap_1(
+// CIR: cir.atomic.xchg seq_cst syncscope(system) %{{.+}}, %{{.+}} : (!cir.ptr<!s8i>, !s8i) -> !s8i
+// LLVM-LABEL: @test_sync_swap_1(
+// LLVM: atomicrmw xchg ptr %{{.+}}, i8 %{{.+}} seq_cst, align 1
+char test_sync_swap_1(char *p, char val) {
+  return __sync_swap(p, val);
+}
+
+// CIR-LABEL: @test_sync_swap_2(
+// CIR: cir.atomic.xchg seq_cst syncscope(system) %{{.+}}, %{{.+}} : (!cir.ptr<!s16i>, !s16i) -> !s16i
+// LLVM-LABEL: @test_sync_swap_2(
+// LLVM: atomicrmw xchg ptr %{{.+}}, i16 %{{.+}} seq_cst, align 2
+short test_sync_swap_2(short *p, short val) {
+  return __sync_swap(p, val);
+}
+
+// CIR-LABEL: @test_sync_swap_4(
+// CIR: cir.atomic.xchg seq_cst syncscope(system) %{{.+}}, %{{.+}} : (!cir.ptr<!s32i>, !s32i) -> !s32i
+// LLVM-LABEL: @test_sync_swap_4(
+// LLVM: atomicrmw xchg ptr %{{.+}}, i32 %{{.+}} seq_cst, align 4
+int test_sync_swap_4(int *p, int val) {
+  return __sync_swap(p, val);
+}
+
+// CIR-LABEL: @test_sync_swap_8(
+// CIR: cir.atomic.xchg seq_cst syncscope(system) %{{.+}}, %{{.+}} : (!cir.ptr<!s64i>, !s64i) -> !s64i
+// LLVM-LABEL: @test_sync_swap_8(
+// LLVM: atomicrmw xchg ptr %{{.+}}, i64 %{{.+}} seq_cst, align 8
+long long test_sync_swap_8(long long *p, long long val) {
+  return __sync_swap(p, val);
+}
+
+// __sync_lock_test_and_set
+
+// CIR-LABEL: @test_sync_lock_test_and_set_1(
+// CIR: cir.atomic.xchg seq_cst syncscope(system) %{{.+}}, %{{.+}} : (!cir.ptr<!s8i>, !s8i) -> !s8i
+// LLVM-LABEL: @test_sync_lock_test_and_set_1(
+// LLVM: atomicrmw xchg ptr %{{.+}}, i8 %{{.+}} seq_cst, align 1
+char test_sync_lock_test_and_set_1(char *p, char val) {
+  return __sync_lock_test_and_set(p, val);
+}
+
+// CIR-LABEL: @test_sync_lock_test_and_set_2(
+// CIR: cir.atomic.xchg seq_cst syncscope(system) %{{.+}}, %{{.+}} : (!cir.ptr<!s16i>, !s16i) -> !s16i
+// LLVM-LABEL: @test_sync_lock_test_and_set_2(
+// LLVM: atomicrmw xchg ptr %{{.+}}, i16 %{{.+}} seq_cst, align 2
+short test_sync_lock_test_and_set_2(short *p, short val) {
+  return __sync_lock_test_and_set(p, val);
+}
+
+// CIR-LABEL: @test_sync_lock_test_and_set_4(
+// CIR: cir.atomic.xchg seq_cst syncscope(system) %{{.+}}, %{{.+}} : (!cir.ptr<!s32i>, !s32i) -> !s32i
+// LLVM-LABEL: @test_sync_lock_test_and_set_4(
+// LLVM: atomicrmw xchg ptr %{{.+}}, i32 %{{.+}} seq_cst, align 4
+int test_sync_lock_test_and_set_4(int *p, int val) {
+  return __sync_lock_test_and_set(p, val);
+}
+
+// CIR-LABEL: @test_sync_lock_test_and_set_8(
+// CIR: cir.atomic.xchg seq_cst syncscope(system) %{{.+}}, %{{.+}} : (!cir.ptr<!s64i>, !s64i) -> !s64i
+// LLVM-LABEL: @test_sync_lock_test_and_set_8(
+// LLVM: atomicrmw xchg ptr %{{.+}}, i64 %{{.+}} seq_cst, align 8
+long long test_sync_lock_test_and_set_8(long long *p, long long val) {
+  return __sync_lock_test_and_set(p, val);
+}
+
+// __sync_lock_release
+
+// CIR-LABEL: @test_sync_lock_release_1(
+// CIR: cir.store {{.*}}atomic(release) {{.*}} : !s8i, !cir.ptr<!s8i>
+// LLVM-LABEL: @test_sync_lock_release_1(
+// LLVM: store atomic i8 0, ptr %{{.+}} release, align 1
+void test_sync_lock_release_1(char *p) {
+  __sync_lock_release(p);
+}
+
+// CIR-LABEL: @test_sync_lock_release_2(
+// CIR: cir.store {{.*}}atomic(release) {{.*}} : !s16i, !cir.ptr<!s16i>
+// LLVM-LABEL: @test_sync_lock_release_2(
+// LLVM: store atomic i16 0, ptr %{{.+}} release, align 2
+void test_sync_lock_release_2(short *p) {
+  __sync_lock_release(p);
+}
+
+// CIR-LABEL: @test_sync_lock_release_4(
+// CIR: cir.store {{.*}}atomic(release) {{.*}} : !s32i, !cir.ptr<!s32i>
+// LLVM-LABEL: @test_sync_lock_release_4(
+// LLVM: store atomic i32 0, ptr %{{.+}} release, align 4
+void test_sync_lock_release_4(int *p) {
+  __sync_lock_release(p);
+}
+
+// CIR-LABEL: @test_sync_lock_release_8(
+// CIR: cir.store {{.*}}atomic(release) {{.*}} : !s64i, !cir.ptr<!s64i>
+// LLVM-LABEL: @test_sync_lock_release_8(
+// LLVM: store atomic i64 0, ptr %{{.+}} release, align 8
+void test_sync_lock_release_8(long long *p) {
+  __sync_lock_release(p);
+}
+
+}

--- a/clang/test/CIR/CodeGenBuiltins/builtin-undef-rvalue.cpp
+++ b/clang/test/CIR/CodeGenBuiltins/builtin-undef-rvalue.cpp
@@ -6,7 +6,7 @@
 typedef int v4si __attribute__((vector_size(16)));
 
 int test_builtin_reduce_add_undef_rvalue(v4si x) {
-  // expected-error@+1 {{unimplemented X86 builtin call: __builtin_reduce_add}}
+  // expected-error@+1 {{unimplemented builtin call: __builtin_reduce_add}}
   return __builtin_reduce_add(x);
 }
 

